BootloaderCorePkg/Library/SmbiosInitLib/SmbiosInitLib.c

```diff
@@ -0,0 +1,389 @@
+/** @file
+
+  Copyright (c) 2017 - 2019, Intel Corporation. All rights reserved.<BR>
+  This program and the accompanying materials
+  are licensed and made available under the terms and conditions of the BSD License
+  which accompanies this distribution.  The full text of the license may be found at
+  http://opensource.org/licenses/bsd-license.php.
+
+  THE PROGRAM IS DISTRIBUTED UNDER THE BSD LICENSE ON AN "AS IS" BASIS,
+  WITHOUT WARRANTIES OR REPRESENTATIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED.
+
+**/
+
+#include <IndustryStandard/SmBios.h>
+#include <Library/MemoryAllocationLib.h>
+#include <Library/DebugLib.h>
+#include <Library/BaseLib.h>
+#include <Library/BaseMemoryLib.h>
+#include <Library/BootloaderCoreLib.h>
+#include <Library/SmbiosInitLib.h>
+#include <Library/PcdLib.h>
+#include <Library/BootloaderCommonLib.h>
+#include "SmbiosTables.h"
+
+VOID    *mType127Ptr      = NULL;
+
+
+/**
+  Check if the Smbios types (including the entry point struct)
+  have crossed the statically allocated size for Smbios init
+
+**/
+EFI_STATUS
+EFIAPI
+CheckSmbiosOverflow (
+  VOID
+  )
+{
+  SMBIOS_TABLE_ENTRY_POINT      *SmbiosEntry;
+
+  SmbiosEntry = (SMBIOS_TABLE_ENTRY_POINT *) PcdGet32 (PcdSmbiosTablesBase);
+
+  if ( (UINT32)(SmbiosEntry->EntryPointLength + sizeof (UINT8) + SmbiosEntry->TableLength) > (UINT32)PcdGet16(PcdSmbiosTablesSize) ) {
+    return TRUE;
+  } else {
+    return FALSE;
+  }
+}
+
+/**
+  Return the pointer to the Smbios table spcified by 'Type'
+
+  @param[in]  Type    Smbios type requested
+
+  @retval             Pointer to the starting of the requested
+                      Smbios type
+
+**/
+VOID *
+FindSmbiosType (
+  IN  UINT8   Type
+  )
+{
+  SMBIOS_STRUCTURE              *TypeHdr;
+  SMBIOS_TABLE_ENTRY_POINT      *SmbiosEntry;
+  UINT8                         *StringPtr;
+
+  SmbiosEntry = (SMBIOS_TABLE_ENTRY_POINT *) PcdGet32 (PcdSmbiosTablesBase);
+
+  if (SmbiosEntry == NULL) {
+    return NULL;
+  }
+
+  TypeHdr = (SMBIOS_STRUCTURE *)(SmbiosEntry->TableAddress);
+  for (; TypeHdr->Type != SMBIOS_TYPE_END_OF_TABLE;) {
```

Comment:
in the loop TypeHdr should be checked against the allocated SmbiosTable boundary.   
EX:   If for some reason, the TYPE_END_OF_TABLE is not appended ,  calling into this function might cause undesired behavior.

Reply:
Added the check here.